### PR TITLE
[MIST-372] Custom formatter and default setup function

### PR DESCRIPTION
**Pull request by rsterbin**
[Review on Reviewable](https://reviewable.io/reviews/mistifyio/mistify-logrus-ext/1)

### CODE REVIEW

**Comment by bbickerton** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

<sup>**[ext.go, line 7 [r1]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-Jk4cVn6f86H4GZbGNYp)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/5c8fe9e2098d4598a32e57b4032be0517cee2aaf/ext.go#L7)):</sup>
I personally think plain `Setup` (or `Configure`) fits better as a function name, but am ok with `DefaultSetup`

---

<sup>**[formatter.go, line 34 [r1]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-Jk4dslSfWK8t5A9xlph)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/5c8fe9e2098d4598a32e57b4032be0517cee2aaf/formatter.go#L34)):</sup>
I really don't like `.` in key names. It breaks simple dot notation nested lookups. Either use suffix `data[k+"Msg"]` or have a new struct with `Foo struct{Data error, Message string}` and `data[k] = Foo{Data: err, Message: err.Error()}`. I think the latter is cleaner and easier to work with the resulting JSON.

---

**Comment by rsterbin** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

<sup>**[ext.go, line 7 [r1]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-Jk4cVn6f86H4GZbGNYp)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/5c8fe9e2098d4598a32e57b4032be0517cee2aaf/ext.go#L7)):</sup>
I was thinking it made it clearer that this was simply our default setup and some repos might need to use something different, but I don't actually know if that's true.  Thoughts?

---

<sup>**[formatter.go, line 34 [r1]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-Jk4dslSfWK8t5A9xlph)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/5c8fe9e2098d4598a32e57b4032be0517cee2aaf/formatter.go#L34)):</sup>
I like that second idea a lot.  One sec, I'll get a new commit up.

---

**Comment by bbickerton** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

Reviewed files:
- hook.go @ r1

---

<sup>**[ext.go, line 7 [r1]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-Jk4cVn6f86H4GZbGNYp)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/5c8fe9e2098d4598a32e57b4032be0517cee2aaf/ext.go#L7)):</sup>
Fair enough. +1

---

**Comment by bbickerton** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

Reviewed files:
- demo/main.go @ r1
- ext.go @ r1

---

File: formatter.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+type ExtJSONFormatter struct{}
+
+type FieldError struct {
+	Error   error
+	Message string
+}
+
+func (f *ExtJSONFormatter) prefixFieldClashes(data log.Fields) {
+	_, ok := data["time"]
+	if ok {
```
Comment (mmlb) [inline formatter.go]: use `if _, ok := date["time"]; ok {` here and below.

**Comment by mmlb** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

<sup>**[formatter.go, line 19 [r2]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-Jk9cVjsq9-SiMKB8pN3)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/3fd927cbb660b88ffa80eb779347a6637e4e1d10/formatter.go#L19)):</sup>
use `if _, ok := date["time"]; ok {` here and below.

---

**Comment by mmlb** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

So I'm not sure I like the hook idea. Problem is that we are modifying the message when a hook is not meant for that. If we ever ship the message to multiple locations I'm not sure how logrus orders the hook calls and we may get a message with missing error.

Seeing as we aren't shipping messages to more than one service and have no immediate plans to do so, I guess I'm ok with it for now as long as we document it in code. But it still doesn't feel right, and I kind of want to go back to the formatter.

---

**Comment by bbickerton** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

Reviewed files:
- ext.go @ r3
- formatter.go @ r3
- formatter_test.go @ r3
- hook.go @ r3
- hook_test.go @ r3

---

Looking at the logrus docs and listed existing hooks, it does seem they're all just message routing and external service notifying. However, the limit of what hooks should be used for and whether they shouldn't modify the message isn't documented. In https://github.com/Sirupsen/logrus/blob/master/hooks.go , it does seem that hooks will be fired in order of addition for each level. As long as it's added first, I think we'll be ok.

---

**Comment by rsterbin** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

My vote's for the hook, because we wouldn't need to maintain our own copy of `prefixFieldClashes`, which is something I could see changing over time (e.g., if they were to come to an agreement on the PR that adds the file and line number).

I see why modifying an entry in a hook might feel wrong, but to me maintaining a duplicate-except-for-this-one-thing formatter feels more wrong.  I think we're doing something Logrus doesn't want us to do anyway and it's going to feel wrong any way we pick, so we might as well pick the one that's easiest to maintain.

---

**Comment by mmlb** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

The problem is that if logrus ever decides to change the implementation of `hooks` we can be left with more breakage/surprises vs always getting what we say (exacerbated by us not vendoring yet). We can always embed `logrus.JSONFormatter` in `mistifyFormatter` and then do the error detection/injection ourselves and pass that into `logrus.JSONFormatter`

IMO hijacking an unrelated subcomponent is worse than pinning our format since that will _never_ change from underneath us.

But like I said, I don't feel _too_ bad about it (I've done worse things ;) ), and am fine as long as we document it in code.

---

**Comment by mmlb** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

Kind of feels like monkey patching which is definitely not Go. I know @bbickerton is all for monkey patching ;)

---

**Comment by bbickerton** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

+1 on embedding formatter

---

**Comment by mmlb** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

<sup>**[formatter.go, line 21 [r4]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-JkYjw-ol7YeYBHIlnot)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/be07a3976b73c947c24544ef2745f9440c69c796/formatter.go#L21)):</sup>
why not do this

```
type MistifyFormatter struct {
    log.JSONFormatter
}
```

and then

`return f.JSONFormatter.Format(entry)`

---

**Comment by bbickerton** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

Reviewed files:
- ext.go @ r4
- formatter.go @ r4
- formatter_test.go @ r4
- hook.go @ r4
- hook_test.go @ r4

---

**Comment by rsterbin** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

<sup>**[formatter.go, line 21 [r4]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-JkYjw-ol7YeYBHIlnot)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/be07a3976b73c947c24544ef2745f9440c69c796/formatter.go#L21)):</sup>
I like it.  Done.

---

**Comment by rsterbin** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

---

<sup>**[formatter.go, line 21 [r4]](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1#-JkYjw-ol7YeYBHIlnot)** ([raw file](https://github.com/mistifyio/mistify-logrus-ext/blob/be07a3976b73c947c24544ef2745f9440c69c796/formatter.go#L21)):</sup>
Done.

---

**Comment by mmlb** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

Reviewed files:
- formatter.go @ r5

---

**Comment by mmlb** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mistifyio/mistify-logrus-ext/1)

Reviewed files:
- formatter.go @ r5

---

**Pull request merged**